### PR TITLE
Add decimal support for IN predicate

### PR DESCRIPTION
**Pull request by pramodsatya**
Add support for decimals in the IN predicate.
Add filter for HugeInt values using a hash table.

### CODE REVIEW

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 1]
```diff
@@ -316,3 +322,25 @@ TEST_F(InPredicateTest, varbinary) {
   auto result = evaluate<SimpleVector<bool>>(predicate, input);
   assertEqualVectors(makeConstant(true, input->size()), result);
 }
+
+TEST_F(InPredicateTest, shortDecimal) {
+  auto type = DECIMAL(5, 2);
```
Comment (karteekmurthys) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:327]: No need to declare variable. Initialize the type in function call.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 2]
```diff
@@ -316,3 +322,25 @@ TEST_F(InPredicateTest, varbinary) {
...
+  auto shortDecimal =
+      makeNullableShortDecimalFlatVector({10000, -10000, 20000, -20000}, type);
+  auto expected = makeFlatVector<bool>({false, true, false, true});
+
+  auto result =
+      evaluate("c0 IN (-100.00, -200.00)", makeRowVector({shortDecimal}));
```
Comment (karteekmurthys) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:333]: Similar to integers, cover cases with nulls in the list, nulls in columns, dictionary etc.

File: velox/type/Filter.cpp [hunk 3]
```diff
@@ -145,15 +145,23 @@ bool BigintValuesUsingBitmask::testInt64Range(
   return !(min > max_ || max < min_);
 }
 
+template <typename T>
 BigintValuesUsingHashTable::BigintValuesUsingHashTable(
```
Comment (karteekmurthys) [inline velox/type/Filter.cpp:149]: May be rename class to `IntegerHashTableFilter`. Since we support int128 as well. Also suffixing with `Filter` makes it clear that it is a filter built using hash table.

File: velox/type/Filter.h [hunk 4]
```diff
@@ -778,12 +786,17 @@ class BigintValuesUsingHashTable final : public Filter {
   // from Murmur hash
   static constexpr uint64_t M = 0xc6a4a7935bd1e995L;
 
-  const int64_t min_;
-  const int64_t max_;
+  int64_t min_;
```
Comment (karteekmurthys) [inline velox/type/Filter.h:789]: Try to declare it as `T min_` and `T max_`.

File: velox/type/Filter.cpp [hunk 5]
```diff
@@ -145,15 +145,23 @@ bool BigintValuesUsingBitmask::testInt64Range(
...
+    T max,
+    const std::vector<T>& values,
     bool nullAllowed)
-    : Filter(true, nullAllowed, FilterKind::kBigintValuesUsingHashTable),
-      min_(min),
-      max_(max),
-      values_(values) {
+    : Filter(true, nullAllowed, FilterKind::kBigintValuesUsingHashTable) {
+  std::vector<T> hashTable;
+  if constexpr (std::is_same<T, int64_t>::value) {
```
Comment (karteekmurthys) [inline velox/type/Filter.cpp:156]: May be we can avoid `if check` by declaring min_ and max_ using type `T`.

File: velox/type/Filter.h [hunk 6]
```diff
@@ -778,12 +786,17 @@ class BigintValuesUsingHashTable final : public Filter {
...
+  int64_t max_;
   std::vector<int64_t> hashTable_;
   bool containsEmptyMarker_ = false;
   std::vector<int64_t> values_;
   int32_t sizeMask_;
+
+  int128_t minLongDecimal_;
+  int128_t maxLongDecimal_;
+  std::vector<int128_t> hashTableLongDecimal_;
+  std::vector<int128_t> valuesLongDecimal_;
```
Comment (karteekmurthys) [inline velox/type/Filter.h:799]: if declared as `std::vector<T> values_;` we may not need this.

**Review by karteekmurthys — COMMENTED** [review]
@pramodsatya thanks for adding these changes. I did my first parse.

File: velox/functions/prestosql/InPredicate.cpp [hunk 7]
```diff
@@ -86,6 +86,120 @@ std::pair<std::unique_ptr<common::Filter>, bool> createBigintValuesFilter(
...
+  }
+  VELOX_USER_CHECK(
+      !values.empty(),
+      "IN predicate expects at least one non-null value in the in-list");
+
+  return {common::createBigintValues(values, nullAllowed), false};
+}
+
+std::optional<std::pair<std::vector<int128_t>, bool>>
+toUnscaledLongDecimalValues(
```
Comment (karteekmurthys) [inline velox/functions/prestosql/InPredicate.cpp:147]: `toUnscaledLongDecimalValues` and `toUnscaledShortDecimalValues` except for int128_t to int64_t is there any other difference? If not, combine them into a function template.

File: velox/functions/prestosql/InPredicate.cpp [hunk 8]
```diff
@@ -86,6 +86,120 @@ std::pair<std::unique_ptr<common::Filter>, bool> createBigintValuesFilter(
...
+  bool nullAllowed = valuesPair.value().second;
+
+  if (values.empty() && nullAllowed) {
+    return {nullptr, true};
+  }
+  VELOX_USER_CHECK(
+      !values.empty(),
+      "IN predicate expects at least one non-null value in the in-list");
+
+  return {common::createHugeintValues(values, nullAllowed), false};
```
Comment (karteekmurthys) [inline velox/functions/prestosql/InPredicate.cpp:200]: Same as above,  this method seems exactly same as the other. Try to combine and minimize code.

File: velox/functions/prestosql/tests/InPredicateTest.cpp (hunk 2 above)
Comment (pramodsatya) [reply velox/functions/prestosql/tests/InPredicateTest.cpp]: Thanks for the suggestion, updated.

File: velox/type/Filter.cpp (hunk 3 above)
Comment (pramodsatya) [reply velox/type/Filter.cpp]: Thanks for the suggestion, renamed to ```Int128ValuesUsingHashTable``` to retain consistency with the current filter naming convention.

File: velox/type/Filter.cpp (hunk 5 above)
Comment (pramodsatya) [reply velox/type/Filter.cpp]: Thanks for the suggestion, templating this class would require changes in other classes as well because of functions like ```mergeWith``` and these changes seem tangential to the current PR. Would it be fine if the filter refactoring is taken up in a subsequent PR?

File: velox/functions/prestosql/InPredicate.cpp (hunk 7 above)
Comment (pramodsatya) [reply velox/functions/prestosql/InPredicate.cpp]: Updated, thanks!

File: velox/type/Filter.cpp [hunk 9]
```diff
@@ -302,6 +305,69 @@ bool BigintValuesUsingHashTable::testInt64Range(
...
+    : Filter(true, nullAllowed, FilterKind::kBigintValuesUsingHashTable),
+      min_(min),
+      max_(max),
+      values_(values) {
+  constexpr int32_t kPaddingElements = 4;
+  VELOX_CHECK(min <= max, "min must be less than max");
+  VELOX_CHECK(!values.empty(), "values must not be empty");
+
+  // Size the hash table to be 2+x the entry count, e.g. 10 entries
+  // gets 1 << log2 of 50 == 32. The filter is expected to fail often so we
```
Comment (karteekmurthys) [inline velox/type/Filter.cpp:322]: nit: `1 << log2 ( 5 * 10)`

File: velox/vector/VariantToVector.cpp [hunk 10]
```diff
@@ -89,6 +111,12 @@ ArrayVectorPtr variantArrayToVector(
   }
 
   const auto elementType = variantArray.at(nonNullIndex.value()).inferType();
+  if (elementType->kind() == TypeKind::SHORT_DECIMAL) {
+    return variantArrayToVectorImpl<TypeKind::SHORT_DECIMAL>(
+        variantArray, pool);
+  } else if (elementType->kind() == TypeKind::LONG_DECIMAL) {
+    return variantArrayToVectorImpl<TypeKind::LONG_DECIMAL>(variantArray, pool);
+  }
```
Comment (karteekmurthys) [inline velox/vector/VariantToVector.cpp:119]: It's not possible to use `VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL`?

File: velox/type/Filter.cpp [hunk 11]
```diff
@@ -302,6 +305,69 @@ bool BigintValuesUsingHashTable::testInt64Range(
...
+}
+
+bool Int128ValuesUsingHashTable::testInt128(int128_t value) const {
+  if (containsEmptyMarker_ && value == kEmptyMarker) {
+    return true;
+  }
+  if (value < min_ || value > max_) {
+    return false;
+  }
+  uint32_t pos = (value * M) & sizeMask_;
```
Comment (karteekmurthys) [inline velox/type/Filter.cpp:357]: This is a 64-bit value, is it ok to use for 128-bit values as well. See if there is an 128-bit equivalent for murmur hash.
```
 // from Murmur hash
  static constexpr uint64_t M = 0xc6a4a7935bd1e995L;
```

**Review by karteekmurthys — COMMENTED** [review]
Left some comments.

File: velox/functions/prestosql/InPredicate.cpp [hunk 12]
```diff
@@ -219,6 +300,22 @@ class InPredicate : public exec::VectorFunction {
...
+            .integerVariable("a_scale")
+            .argumentType("DECIMAL(a_precision, a_scale)")
+            .argumentType("array(DECIMAL(a_precision, a_scale))")
+            .build());
+    signatures.emplace_back(exec::FunctionSignatureBuilder()
+                                .returnType("boolean")
+                                .integerVariable("a_precision")
+                                .integerVariable("a_scale")
+                                .argumentType("DECIMAL(a_precision, a_scale)")
+                                .argumentType("array(unknown)")
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/InPredicate.cpp:316]: This is required for IN-list with only null values, such as in the test ```c0 IN (null)```.

File: velox/functions/prestosql/InPredicate.cpp [hunk 13]
```diff
@@ -86,6 +86,67 @@ std::pair<std::unique_ptr<common::Filter>, bool> createBigintValuesFilter(
   return {common::createBigintValues(values, nullAllowed), false};
 }
 
+template <typename T, typename U>
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/InPredicate.cpp:89]: Thanks for the suggestion, this could be used to initialize ```values```. But the typename for int is required for the function return value.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 14]
```diff
@@ -166,6 +172,92 @@ class InPredicateTest : public FunctionBaseTest {
...
+  template <typename T>
+  void testDecimals(
+      const std::vector<std::vector<std::optional<T>>> unscaledValues,
+      const TypePtr& type) {
+    const auto size = 5;
+    RowVectorPtr rowVector = nullptr;
+    if constexpr (std::is_same_v<T, int128_t>) {
+      rowVector = makeRowVector(
+          {makeNullableLongDecimalFlatVector(unscaledValues[0], type),
+           makeNullableLongDecimalFlatVector(unscaledValues[1], type)});
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/tests/InPredicateTest.cpp:185]: Thanks for the suggestion but ```makeNullableFlatVector``` does not support UnscaledShortDecimal and UnscaledLongDecimal types.

File: velox/vector/VariantToVector.cpp (hunk 10 above)
Comment (pramodsatya) [reply velox/vector/VariantToVector.cpp:119]: No this hadn't worked before. I will rebase this on top of the merged changes by Masha and revert all current changes to this file.

File: velox/type/Filter.cpp (hunk 11 above)
Comment (pramodsatya) [reply velox/type/Filter.cpp]: Thanks for the suggestion, the unit tests worked with the 64 bit value but replaced with the correct Int128 value from MurmurHash3.

File: velox/functions/prestosql/InPredicate.cpp [hunk 15]
```diff
@@ -59,10 +59,10 @@ std::optional<std::pair<std::vector<T>, bool>> toValues(
 // Creates a filter for constant values. A null filter means either
 // no values or only null values. The boolean is true if the list is
 // non-empty and consists of nulls only.
-template <typename T>
-std::pair<std::unique_ptr<common::Filter>, bool> createBigintValuesFilter(
+template <typename T, typename U>
+std::pair<std::unique_ptr<common::Filter>, bool> createValuesFilter(
```
Comment (majetideepak) [inline velox/functions/prestosql/InPredicate.cpp:63]: We are gaining any reuse by changing `createBigintValuesFilter` to `createValuesFilter` since we specialize `int128_t` below.
Let's keep `createBigintValuesFilter` as is and add `createHugeintValuesFilter`

Comment (pramodsatya) [reply velox/functions/prestosql/InPredicate.cpp]: Thanks for the suggestion, updated.

**Comment by majetideepak** [comment]
@karteekmurthys have you complete your review of this PR?

**Comment by karteekmurthys** [comment]
@majetideepak looks good. @pramodsatya please update description.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 16]
```diff
@@ -167,6 +167,92 @@ class InPredicateTest : public FunctionBaseTest {
...
+      rowVector = makeRowVector(
+          {makeNullableShortDecimalFlatVector(unscaledValues[0], type),
+           makeNullableShortDecimalFlatVector(unscaledValues[1], type)});
+    }
+
+    // IN list with decimals and without nulls.
+    auto expected = makeFlatVector<bool>(
+        size, [](auto row) { return row == 0 || row == 3; });
+    VectorPtr result;
+    if constexpr (std::is_same_v<T, int128_t>) {
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/tests/InPredicateTest.cpp]: Thanks for the suggestion, moved decimal tests to a separate class and separated into short decimal and long decimal test utility functions for readability.

**Review by majetideepak — COMMENTED** [review]
I see there is scope for reuse between `HugeintValuesUsingHashTable` and `BigintValuesUsingHashTable`. How about creating an `IntValuesUsingHashTable` base class and moving the common pieces there?

File: velox/type/Filter.cpp [hunk 17]
```diff
@@ -918,6 +935,115 @@ bool BigintValuesUsingHashTable::testInt64Range(
...
+    if (l == value) {
+      return true;
+    }
+  }
+  return false;
+}
+
+folly::dynamic HugeintValuesUsingHashTable::serialize() const {
+  auto obj = Filter::serializeBase("HugeintValuesUsingHashTable");
+  obj["min_lower"] = HugeInt::lower(min_);
```
Comment (majetideepak) [inline velox/type/Filter.cpp:999]: Can't we serialize int128_t values?

Comment (karteekmurthys) [reply velox/type/Filter.cpp:969]: +1

Comment (pramodsatya) [reply velox/type/Filter.cpp:969]: Thanks for the suggestion, I checked ```asImpl() in folly/dynamic-inl.h``` before and could not find a function to serialize int128 values. Will check if there are other folly functions to serialize int128 values and update here.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 18]
```diff
@@ -328,3 +328,156 @@ TEST_F(InPredicateTest, reusableResult) {
   auto expected = makeFlatVector<bool>({false, true, true, false});
   assertEqualVectors(expected, actual);
 }
+
+class InPredicateDecimalTest : public FunctionBaseTest {
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/functions/prestosql/tests/InPredicateTest.cpp:332]: The values in the IN-list are hardcoded and the input vector is constructed inside the function so they cannot be re-used currently. A common test utility function with a template variable can be used as attempted before for short and long decimals, but it could affect the readability.
Please let me know if I should use a common function.

File: velox/type/Filter.h [hunk 19]
```diff
@@ -932,6 +933,70 @@ class BigintValuesUsingHashTable final : public Filter {
...
+    }
+  }
+
+  const std::vector<int128_t>& values() const {
+    return values_;
+  }
+
+  const std::vector<int128_t>& hashTable() const {
+    return hashTable_;
+  }
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/type/Filter.h:986]: Thanks for the suggestion, updated.

File: velox/vector/VariantToVector.cpp [hunk 20]
```diff
@@ -69,7 +77,7 @@ ArrayVectorPtr variantArrayToVector(
         arrayType, variantArray, pool);
   }
 
-  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
```
Comment (majetideepak) [inline velox/vector/VariantToVector.cpp:72]: `VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH` now covers BIGINT and HUGEINT. We don't need the changes in this file.

File: velox/type/Filter.h [hunk 21]
```diff
@@ -865,6 +868,100 @@ class BigintValuesUsingHashTable final : public Filter {
...
+ protected:
+  const T min_;
+  const T max_;
+  std::vector<T> hashTable_;
+  bool containsEmptyMarker_ = false;
+  std::vector<T> values_;
+  int32_t sizeMask_;
+
+  // Empty marker for the int64_t hash table.
+  static constexpr int64_t kBigintEmptyMarker = 0xdeadbeefbadefeedL;
```
Comment (majetideepak) [inline velox/type/Filter.h:922]: Can these be moved to the corresponding sub-classes?

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 22]
```diff
@@ -22,18 +22,34 @@ using namespace facebook::velox::functions::test;
 class InPredicateTest : public FunctionBaseTest {
  protected:
   template <typename T>
-  void testIntegers() {
+  void testValues(const TypePtr type) {
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:25]: add a default type based on the physical type from TypeTraits of T.

**Review by majetideepak — COMMENTED** [review]
@pramodsatya I made a high-level pass. Found some improvements.
The filter class looks much cleaner. 
However, the test needs improvement for the inList.
Can we keep the inList same and change the input data for decimals?

File: velox/type/Filter.h (hunk 21 above)
Comment (pramodsatya) [reply velox/type/Filter.h:922]: Thanks for the suggestion.. but because the hashtable construction and insertion of values from the inList for both ```BigintValuesUsingHashTable``` and ```HugeintValuesUsingHashTable``` is happening in the base class, ```IntValuesUsingHashTable```, it cannot be moved without duplicating this code.

File: velox/vector/VariantToVector.cpp (hunk 20 above)
Comment (pramodsatya) [reply velox/vector/VariantToVector.cpp]: Updated, thanks!

**Comment by pramodsatya** [comment]
Thanks for the review @majetideepak. The precision and scale of values in the inList should be the same as that of the input. The inList is hence modified depending on the datatype used in the tests (DECIMAL(5, 2), DECIMAL(19, 7)). Could you please share if you have any inputs on how to do this?

**Comment by majetideepak** [comment]
> The precision and scale of values in the inList should be the same as that of the input

I see. The current if block for each inList is reducing readability.

**Comment by pramodsatya** [comment]
> > The precision and scale of values in the inList should be the same as that of the input
> 
> I see. The current if block for each inList is reducing readability.

Thanks for the note @majetideepak, I have modified the tests a bit to improve the readability. Could you please review the changes?

File: velox/type/Filter.cpp [hunk 23]
```diff
@@ -143,6 +146,18 @@ std::vector<int64_t> deserializeValues(const folly::dynamic& obj) {
   }
   return values;
 }
+
+std::vector<int128_t> deserializeHugeintValues(const folly::dynamic& obj) {
+  std::vector<int128_t> values;
+  auto lowerValuesArray = obj["lower_values"];
+  auto upperValuesArray = obj["upper_values"];
+  auto len = lowerValuesArray.size();
+  values.reserve(len);
```
Comment (majetideepak) [inline velox/type/Filter.cpp:155]: `values.resize(len)`;

File: velox/type/Filter.cpp [hunk 24]
```diff
@@ -143,6 +146,18 @@ std::vector<int64_t> deserializeValues(const folly::dynamic& obj) {
...
+  for (auto i = 0; i < len; i++) {
+    HugeInt::build(upperValuesArray[i].asInt(), lowerValuesArray[i].asInt());
```
Comment (majetideepak) [inline velox/type/Filter.cpp:157]: Shouldn't this be `values[i] = HugeInt::build(upperValuesArray[i].asInt(), lowerValuesArray[i].asInt());`

Can we add a test?

File: velox/type/Filter.cpp [hunk 25]
```diff
@@ -918,6 +920,56 @@ bool BigintValuesUsingHashTable::testInt64Range(
...
+  auto nullAllowed = deserializeNullAllowed(obj);
+  auto min = HugeInt::build(obj["min_upper"].asInt(), obj["min_lower"].asInt());
+  auto max = HugeInt::build(obj["max_upper"].asInt(), obj["max_lower"].asInt());
+  auto values = deserializeHugeintValues(obj);
+
+  return std::make_unique<HugeintValuesUsingHashTable>(
+      min, max, values, nullAllowed);
+}
+
+bool HugeintValuesUsingHashTable::testingEquals(const Filter& other) const {
```
Comment (majetideepak) [inline velox/type/Filter.cpp:952]: Can we template and re-use for BigInt and HugeInt?
Probably use `std::is_same_v` for the cast?

File: velox/type/DecimalUtil.cpp [hunk 26]
```diff
@@ -19,7 +19,11 @@
 
 namespace facebook::velox {
 namespace {
-std::string formatDecimal(uint8_t scale, int128_t unscaledValue) {
+std::string formatDecimal(
```
Comment (majetideepak) [inline velox/type/DecimalUtil.cpp:22]: Can we move this to a separate PR with tests?

**Review by majetideepak — COMMENTED** [review]
@pramodsatya some more comments. The `getInList<T>` is a nice approach. But it can be simplified further. I will make a pass.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 27]
```diff
@@ -22,18 +22,56 @@ using namespace facebook::velox::functions::test;
...
+      return DecimalUtil::toString(flatVec->valueAt(idx), type, true);
+    }
+    return std::to_string(flatVec->valueAt(idx));
+  }
+
+  template <typename T>
+  std::string getInList(
+      std::vector<std::optional<T>> input,
+      const TypePtr& type = CppToType<T>::create()) {
+    FlatVectorPtr<T> flatVec = makeNullableFlatVector<T>(input, type);
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:42]: What does flatVec->toString() give?

File: velox/type/Filter.cpp (hunk 25 above)
Comment (pramodsatya) [reply velox/type/Filter.cpp]: Updated, thanks!

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 28]
```diff
@@ -22,18 +22,51 @@ using namespace facebook::velox::functions::test;
...
+  void testValues(const TypePtr type = CppToType<T>::create()) {
+    if (type->isDecimal()) {
+      this->options_.parseDecimalAsDouble = false;
+    }
     std::shared_ptr<memory::MemoryPool> pool{
         memory::addDefaultLeafMemoryPool()};
 
     const vector_size_t size = 1'000;
-    auto vector = makeFlatVector<T>(size, [](auto row) { return row % 17; });
+    std::string inList;
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:58]: nit: remove this variable and use `auto inList` below for the first use.

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 29]
```diff
@@ -141,52 +184,71 @@ class InPredicateTest : public FunctionBaseTest {
...
-  void testsIntegerConstant() {
+  void testConstantValues(const TypePtr type = CppToType<T>::create()) {
     const vector_size_t size = 1'000;
-    auto rowVector = makeRowVector(
-        {makeConstant((T)123, size),
-         makeNullConstant(CppToType<T>::create()->kind(), size)});
+    VectorPtr c1 = type->isDecimal()
+        ? BaseVector::createNullConstant(type, size, pool())
+        : makeNullConstant(CppToType<T>::create()->kind(), size);
+    auto rowVector = makeRowVector({makeConstant((T)123, size, type), c1});
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:192]: `auto rowVector = makeRowVector({makeConstant(static_cast<T>(123), size, type), BaseVector::createNullConstant(type, size, pool())});`

File: velox/functions/prestosql/tests/InPredicateTest.cpp [hunk 30]
```diff
@@ -141,52 +184,71 @@ class InPredicateTest : public FunctionBaseTest {
   }
 
   template <typename T>
-  void testsIntegerConstant() {
+  void testConstantValues(const TypePtr type = CppToType<T>::create()) {
     const vector_size_t size = 1'000;
-    auto rowVector = makeRowVector(
-        {makeConstant((T)123, size),
-         makeNullConstant(CppToType<T>::create()->kind(), size)});
+    VectorPtr c1 = type->isDecimal()
```
Comment (majetideepak) [inline velox/functions/prestosql/tests/InPredicateTest.cpp:189]: remove c1 and inline as before using `BaseVector::createNullConstant`

File: velox/type/Filter.h [hunk 31]
```diff
@@ -865,6 +868,102 @@ class BigintValuesUsingHashTable final : public Filter {
...
+
+  const std::vector<T>& values() const {
+    return values_;
+  }
+
+  const std::vector<T>& hashTable() const {
+    return hashTable_;
+  }
+
+  bool testInt(T value) const {
```
Comment (majetideepak) [inline velox/type/Filter.h:887]: move this to cpp file

File: velox/type/Filter.h [hunk 32]
```diff
@@ -865,6 +868,102 @@ class BigintValuesUsingHashTable final : public Filter {
...
+  // https://github.com/apache/commons-codec/blob/master/src/main/java/org/apache/commons/codec/digest/MurmurHash3.java
+  static constexpr int128_t kHugeintMurmurHash =
+      HugeInt::build(0x87c37b91114253d5L, 0x4cf5ad432745937fL);
+};
+
+/// IN-list filter for int64_t data type, implemented as a hash table.
+class BigintValuesUsingHashTable final
+    : public IntValuesUsingHashTable<int64_t> {
+ public:
+  BigintValuesUsingHashTable(
```
Comment (majetideepak) [inline velox/type/Filter.h:941]: move to CPP file.

**Review by majetideepak — COMMENTED** [review]
@pramodsatya some more comments. Looks much better now.

File: velox/type/Filter.h [hunk 33]
```diff
@@ -913,22 +998,66 @@ class BigintValuesUsingHashTable final : public Filter {
...
+};
 
-  static constexpr int64_t kEmptyMarker = 0xdeadbeefbadefeedL;
-  // from Murmur hash
-  static constexpr uint64_t M = 0xc6a4a7935bd1e995L;
+/// IN-list filter for int128_t data type, implemented as a hash table.
+class HugeintValuesUsingHashTable final
+    : public IntValuesUsingHashTable<int128_t> {
+ public:
+  HugeintValuesUsingHashTable(
```
Comment (majetideepak) [inline velox/type/Filter.h:1014]: move the definition to cpp file.

File: velox/type/Filter.h (hunk 31 above)
Comment (pramodsatya) [reply velox/type/Filter.h:887]: Thanks for the suggestion, this was in the cpp file earlier but it resulted in the linux-build and linux-adapters CI jobs to fail with function not found error. Moving the definition to the header fixed the error.

Comment (majetideepak) [reply velox/type/Filter.h:887]: You need to define testInt128, and testInt64 in the CPP file as well.

**Comment by majetideepak** [comment]
@pramodsatya can you rebase this PR with the main branch?

**Comment by pramodsatya** [comment]
Thanks for the review @majetideepak, the CI error is fixed now.

File: velox/type/Filter.h [hunk 34]
```diff
@@ -925,22 +968,48 @@ class BigintValuesUsingHashTable final : public Filter {
...
   std::unique_ptr<Filter>
   mergeWith(int64_t min, int64_t max, const Filter* other) const;
+};
 
-  static constexpr int64_t kEmptyMarker = 0xdeadbeefbadefeedL;
-  // from Murmur hash
-  static constexpr uint64_t M = 0xc6a4a7935bd1e995L;
+/// IN-list filter for int128_t data type, implemented as a hash table.
+class HugeintValuesUsingHashTable final
+    : public IntValuesUsingHashTable<int128_t> {
```
Comment (Yuhta) [inline velox/type/Filter.h:982]: I don't think you need to inherit `IntValuesUsingHashTable` here.  Just keep a `F14FastSet<int128_t>` and use it as hash table.  The point of hash table in `BigintValuesUsingHashTable` is to do `testValues`, which is not very beneficial for wide values like int128.

Comment (pramodsatya) [reply velox/type/Filter.h]: Updated, thanks for the suggestion!
